// framer-motion/dist/es/animation/utils/default-transitions.js
import {
    __assign
} from 'tslib';
import {
    isKeyframesTarget
} from './is-keyframes-target.js';

var underDampedSpring = function underDampedSpring() {
    return {
        type: "spring",
        stiffness: 500,
        damping: 25,
        restDelta: 0.5,
        restSpeed: 10
    };
};

var criticallyDampedSpring = function criticallyDampedSpring(to) {
    return {
        type: "spring",
        stiffness: 550,
        damping: to === 0 ? 2 * Math.sqrt(550) : 30,
        restDelta: 0.01,
        restSpeed: 10
    };
};

var linearTween = function linearTween() {
    return {
        type: "keyframes",
        ease: "linear",
        duration: 0.3
    };
};

var keyframes = function keyframes(values) {
    return {
        type: "keyframes",
        duration: 0.8,
        values: values
    };
};

var defaultTransitions = {
    x: underDampedSpring,
    y: underDampedSpring,
    z: underDampedSpring,
    rotate: underDampedSpring,
    rotateX: underDampedSpring,
    rotateY: underDampedSpring,
    rotateZ: underDampedSpring,
    scaleX: criticallyDampedSpring,
    scaleY: criticallyDampedSpring,
    scale: criticallyDampedSpring,
    opacity: linearTween,
    backgroundColor: linearTween,
    color: linearTween,
    default: criticallyDampedSpring
};

var getDefaultTransition = function getDefaultTransition(valueKey, to) {
    var transitionFactory;

    if (isKeyframesTarget(to)) {
        transitionFactory = keyframes;
    } else {
        transitionFactory = defaultTransitions[valueKey] || defaultTransitions.default;
    }

    return __assign({
        to: to
    }, transitionFactory(to));
};

export {
    criticallyDampedSpring,
    getDefaultTransition,
    linearTween,
    underDampedSpring
};
